File: backend/app/schemas/question.py
import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

QuestionSource = Literal["manual", "ai_generated", "ai_improved"]
QuestionStatus = Literal["draft", "approved", "asked", "answered"]
AnswerStatus = Literal["not_answered", "answered", "uncertain"]
AnswerSource = Literal["manual", "ai_extracted"]


class QuestionBase(BaseModel):
    text: str = Field(min_length=1)
    source: QuestionSource = "manual"
    status: QuestionStatus = "draft"
    topic: str | None = None
    position: int = Field(default=0, ge=0)
    is_important: bool = False
    is_optional: bool = False
    notes: str | None = None


class QuestionCreate(QuestionBase):
    pass


class QuestionUpdate(BaseModel):
    text: str | None = Field(default=None, min_length=1)
    source: QuestionSource | None = None
    status: QuestionStatus | None = None
    topic: str | None = None
    position: int | None = Field(default=None, ge=0)
    is_important: bool | None = None
    is_optional: bool | None = None
    notes: str | None = None


class QuestionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    guest_id: uuid.UUID
    text: str = Field(validation_alias="text_")
    source: QuestionSource
    status: QuestionStatus
    topic: str | None = None
    position: int
    is_important: bool
    is_optional: bool
    notes: str | None = None
    spoken_question: str | None = None
    answer: str | None = None
    answer_status: AnswerStatus
    answer_source: AnswerSource | None = None
    answer_updated_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class QuestionAnswerUpdate(BaseModel):
    """PATCH /api/questions/{question_id}/answer.

    `answer` is required (but may be null/blank to clear the answer) -
    that's the whole point of this endpoint. `spoken_question` is optional
    and only updated if the key is present in the request at all (tracked
    via exclude_unset in the service layer), even if its value is null.
    """

    answer: str | None
    spoken_question: str | None = None

File: frontend/services/questions/types/question.ts
/**
 * Mirrors backend/app/schemas/question.py, question_generation.py,
 * question_improvement.py, and question_version.py exactly.
 */

export type QuestionSource = "manual" | "ai_generated" | "ai_improved";
export type QuestionStatus = "draft" | "approved" | "asked" | "answered";
export type AnswerStatus = "not_answered" | "answered" | "uncertain";
export type AnswerSource = "manual" | "ai_extracted";

export type Question = {
  id: string;
  guest_id: string;
  text: string;
  source: QuestionSource;
  status: QuestionStatus;
  topic: string | null;
  category: string | null;
  priority: string | null;
  intent_summary: string | null;
  position: number;
  is_important: boolean;
  is_optional: boolean;
  notes: string | null;
  research_id: string | null;
  research_version: number | null;
  research_item_ids: string[];
  source_urls: string[];
  follow_up_questions: string[];
  generation_reason: string | null;
  generation_run_id: string | null;
  generation_candidate_id: string | null;
  // Interview-tab fields - present on the response but not surfaced by the
  // Questions workspace (that's the Interview tab's job).
  spoken_question: string | null;
  answer: string | null;
  answer_status: AnswerStatus;
  answer_source: AnswerSource | null;
  answer_updated_at: string | null;
  created_at: string;
  updated_at: string;
};

export type QuestionCreateInput = {
  text: string;
  source?: QuestionSource;
  status?: QuestionStatus;
  topic?: string | null;
  position?: number;
  is_important?: boolean;
  is_optional?: boolean;
  notes?: string | null;
};

export type QuestionUpdateInput = Partial<QuestionCreateInput>;

export type QuestionVersion = {
  id: string;
  question_id: string;
  version: number;
  text: string;
  source: QuestionSource;
  created_at: string;
};

// --- Generation ---

export type QuestionGenerationLanguage = "ar" | "en";
export type QuestionGenerationPriority = "low" | "medium" | "high";

export type QuestionGenerationRequest = {
  count: number;
  language: QuestionGenerationLanguage;
  style: string;
  include_followups: boolean;
  topics?: string[] | null;
};

export type GeneratedQuestion = {
  candidate_id: string;
  text: string;
  topic: string | null;
  category: string;
  intent_summary: string;
  priority: QuestionGenerationPriority;
  research_item_ids: string[];
  source_urls: string[];
  reason: string | null;
  follow_up_questions: string[];
};

export type QuestionGenerationResponse = {
  guest_id: string;
  generation_run_id: string;
  research_id: string;
  research_version: number;
  generator_provider: string;
  generator_model: string | null;
  requested_count: number;
  candidate_count: number;
  generated_count: number;
  duplicates_filtered_count: number;
  refill_attempts: number;
  questions: GeneratedQuestion[];
};

export type SelectedGeneratedQuestion = GeneratedQuestion;

export type QuestionGenerationSaveRequest = {
  generation_run_id: string;
  research_id: string;
  research_version: number;
  questions: GeneratedQuestion[];
};

export type SkippedGeneratedQuestion = {
  candidate_id: string | null;
  text: string;
  reason: "already_saved" | "exact_duplicate" | "semantic_duplicate" | "concurrent_duplicate";
  duplicate_of_question_id: string | null;
};

export type QuestionGenerationSaveResponse = {
  guest_id: string;
  saved_count: number;
  skipped_count: number;
  questions: Question[];
  skipped: SkippedGeneratedQuestion[];
};

// --- Improvement ---

export type QuestionImprovementLanguage = "ar" | "en";
export type QuestionImprovementStyle = "conversational" | "professional" | "deep" | "concise";
export type QuestionImprovementGoal = "clarity" | "depth" | "brevity" | "natural";

export type QuestionImprovementRequest = {
  language?: QuestionImprovementLanguage;
  style?: QuestionImprovementStyle;
  goal?: QuestionImprovementGoal;
};

export type QuestionImprovementPreview = {
  question_id: string;
  original_text: string;
  improved_text: string;
  reason: string | null;
  changes: string[];
};
